service/history/api/respondactivitytaskcompleted/api.go

```diff
@@ -97,13 +100,26 @@ func Invoke(
 			}
 
 			if !isRunning ||
-				ai.StartedEventId == common.EmptyEventID ||
+				(!isCompletedByID && ai.StartedEventId == common.EmptyEventID) ||
 				(token.GetScheduledEventId() != common.EmptyEventID && token.Attempt != ai.Attempt) ||
 				(token.GetVersion() != common.EmptyVersion && token.Version != ai.Version) {
 				return nil, consts.ErrActivityTaskNotFound
 			}
 
-			if _, err := mutableState.AddActivityTaskCompletedEvent(scheduledEventID, ai.StartedEventId, request); err != nil {
+			// We fabricate a started event only when the activity is not started yet and
+			// we need to force complete an activity
+			if isCompletedByID && ai.StartedEventId == common.EmptyEventID {
+				_, err := mutableState.AddActivityTaskStartedEvent(ai, scheduledEventID,
+					"",
+					req.GetCompleteRequest().GetIdentity(),
+					nil)
+				if err != nil {
+					return nil, err
+				}
+			}
+
+			ai, _ = mutableState.GetActivityInfo(scheduledEventID)
+			if _, err = mutableState.AddActivityTaskCompletedEvent(scheduledEventID, ai.StartedEventId, request); err != nil {
 				// Unable to add ActivityTaskCompleted event to history
 				return nil, err
 			}
```

Comment:
The ActivityE2ELatency metric emitted below doesn't make sense in this case as activityStartedTime will be now and latency will be 0. We should skip that emitting logic I think.

Reply:
The discussion activityStartedTime is out of the scope of this PR. If you would like to discuss about it, we'd better to create an issue to track it. 

Thank you.

Reply:
Not quite following why that's out of scope of this PR. Before this PR that metric value is always valid and reflects the E2E latency. With this change, both activity started and completed are done at the same time and the value emitted becomes meaningless.

Reply:
That is part of that, for the normal path, we still need that right?
Beside this, we may treat this the metric value emitted may be used as a metric to track on the force completion cases.

Reply:
yeah. I was only trying to suggest skipping the metric emitting logic in this force completion case in my original comment. Sorry if that message causes any confusion. 

> treat this the metric value emitted may be used as a metric to track on the force completion cases.

We will need a separate metric or metric tag to tell the value is for force completion case when querying the metric. And that's actually good point, we may want to know how many activities get force completed. 
I suggested skipping here because that's the easiest change to make the PR self-contained and we can get it landed.

Reply:
I see, thanks for your clarification.

My concern here is if we skip the metric tracking for the force completion ones now, that means we lose the data points to track the service. Is it OK to lose data points here? If we do not want, I can try to add a separate metric or metric tag.

Reply:
> lose the data points to track the service

Can you elaborate the concern a bit more? The metric to track what service?

Not emitting the activity E2E latency in this case make senses to me because we can't compute the right value anyway (unless we add a startedTime field into the respondActivityCompletedByIdRequest I guess). 
We still have other metrics for this API call though in both frontend and history service to track the # of requests and the request latency. 

We can add an additional metric in the history api handler to track the number of force completions we do. I am ok with doing that in a follow up PR.

Reply:
Agree with @yycptt, lets not record the E2E latency metric if we fabricated the start.

Reply:
I may not have the full view on the metric, but to my thought, let's say we still need to track the number of events we receive, and the number of those events should be some how match the count of the metric we emitted here. However, if the above case is not an issue. I can start working on a new reversion, and will create an issue to create the metric as a follow up PR for this one.

Let me know your thoughts and thank you!

Reply:
> need to track the number of events we receive, and the number of those events should be some how match the count of the metric we emitted here

hmm no we don't have such invariant/guarantee today. It's also hard to do I guess because service can fail anything between two metric emitting point. Metrics system is also not that reliable and can't guarantee the numbers will perfectly match all the time I think. 

Not recoding E2E latency for force completion is perfectly fine here.